### PR TITLE
Add threads per core

### PR DESCRIPTION
**Pull request by rcvalerio**
[#297](https://github.com/inductiva/tasks/issues/297)
Client implementation for https://github.com/inductiva/inductiva-web-api/pull/1190

This PR adds an option to define the number of threads_per_core, default is 2 can be changed to 1. This is the same behavior of the google cloud.

For more info, check:
https://cloud.google.com/compute/docs/instances/set-threads-per-core

How to test: 
```python
import inductiva

mg = inductiva.resources.MachineGroup(machine_type="c2-standard-4",
                                      num_machines=1,
                                      threads_per_core=1)

mg.start()
```
And then check on google cloud for the vCPUs to core ratio of the VM.

Note:  mg.n_vcpus does not correctly display the number of vCPUS. The implementation currently only checks the vCPUs of the machine_type. It does not actually query the machine.

Note2: "In the Google Cloud console, when the number of threads per core is set to 1, the CPU utilization metric can only scale up to a maximum of 50%. When the number of threads per core is set to 2, the CPU utilization metric can scale up to 100%."

### CODE REVIEW

File: inductiva/resources/machines_base.py [hunk 1]
```diff
@@ -175,12 +178,14 @@ def _register_machine_group(self, **kwargs):
...
             machine_type=self.machine_type,
             provider_id=self.provider,
+            threads_per_core=self.threads_per_core,
             disk_size_gb=self.data_disk_gb,
             max_idle_time=self._timedelta_to_seconds(self.max_idle_time),
             auto_terminate_ts=self._convert_auto_terminate_ts(
                 self.auto_terminate_ts),
             **kwargs,
         )
+        print(instance_group_config)
```
Comment (bgregoinductiva) [inline inductiva/resources/machines_base.py:188]: Please remove this debug print.

**Review by sergiomsantos — COMMENTED** [review]
We're actually enabling/disabling SMT, not changing the number of threads per core.
As it is, it gives the impression that the user can set `threads_per_core` to whatever amount he wants. Is this the case? If not, I suggest changing the argument to `disable_simultaneous_multithreading` with a default value of `False`

**Comment by rcvalerio** [comment]
> We're actually enabling/disabling SMT, not changing the number of threads per core. As it is, it gives the impression that the user can set `threads_per_core` to whatever amount he wants. Is this the case? If not, I suggest changing the argument to `disable_simultaneous_multithreading` with a default value of `False`

This was the initial idea, then we decided to change to threads_per_core to keep it consistent with google cloud: https://github.com/inductiva/inductiva-web-api/pull/1190#issuecomment-2332709241
https://cloud.google.com/compute/docs/instances/set-threads-per-core

File: inductiva/resources/machine_cluster.py [hunk 2]
```diff
@@ -18,6 +18,7 @@ def __init__(
         self,
         machine_type: str,
         num_machines: int = 2,
+        threads_per_core: int = 2,
```
[the remark this answers is not in the document]
Comment (rcvalerio) [reply inductiva/resources/machine_cluster.py:21]: Yes, validation on the client was missing. I only had validation on the backend.

File: inductiva/resources/machines_base.py [hunk 3]
```diff
@@ -66,8 +68,12 @@ def __init__(
         if data_disk_gb <= 0:
             raise ValueError("`data_disk_gb` must be positive.")
 
+        if threads_per_core not in [1, 2]:
+            raise ValueError("`threads_per_core` must be either 1 or 2.")
+
```
[the remark this answers is not in the document]
Comment (rcvalerio) [reply inductiva/resources/machines_base.py:73]: Yes, we now have a test for checking the threads_per_core client validation.

**Review by sergiomsantos — APPROVED** [review]

**Review by bgregoinductiva — APPROVED** [review]

**Pull request merged**